### PR TITLE
Install swaybar and swaybg under /usr/lib/sway

### LINKED ISSUE #1880: Install swaybg and swaybar in /usr/lib/sway/
Binaries like `swaybg` and `swaybar` aren't meant to be invoked directly by the user, and so do not belong in `${PATH}`.

The typical way to deal with it, exemplified by `git`, is to use `/usr/lib/${name}` (here, `/usr/lib/sway`), and fallback to searching in `${PATH}` so that executables can also be user-supplied.

### PR DESCRIPTION
**Pull request by nbraud**
Closes #1880

### CODE REVIEW

File: sway/commands/exec_always.c [hunk 1]
```diff
@@ -51,7 +52,36 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
...
 		// Fork child process again
 		setsid();
+
 		if ((*child = fork()) == 0) {
+			// Acquire the current PATH
+			char* path = secure_getenv('PATH');
+			const char* extra_path = ':/usr/lib/sway';
+			const size_t extra_size = sizeof(extra_path);
+
+			if(!path) {
```
Comment (SirCmpwn) [inline sway/commands/exec_always.c]: Add space before the (

File: sway/commands/exec_always.c [hunk 2]
```diff
@@ -51,7 +52,36 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
 	if ((pid = fork()) == 0) {
 		// Fork child process again
 		setsid();
+
 		if ((*child = fork()) == 0) {
+			// Acquire the current PATH
+			char* path = secure_getenv('PATH');
+			const char* extra_path = ':/usr/lib/sway';
```
Comment (SirCmpwn) [inline sway/commands/exec_always.c]: All of these strings need double quotes, did you even try to compile this before submitting it?

And review the style guide, you have numerous errors

**Comment by nbraud** [comment]
I'm not sure whether hardcoding `/usr/lib/sway` and `secure_getenv(3)` is appropriate; will `sway` ever support non-Linux or non-GLibC environments?

**Comment by SirCmpwn** [comment]
It already does, we support musl and FreeBSD. We forbid GNU extensions.

File: sway/commands/exec_always.c (hunk 2 above)
Comment (nbraud) [reply sway/commands/exec_always.c]: Yes, I got confused while amending the patch for style, and didn't rebuild right before resubmitting...

**Comment by Hummer12007** [comment]
Why not /usr/libexec?

**Comment by SirCmpwn** [comment]
>Why not /usr/libexec?

Meh

**Comment by nbraud** [comment]
@Hummer12007 I made `INSTLIBDIR` configurable, so you can make it be `/usr/libexec` or `/usr/local/libexec` on systems which use `libexec`

**Comment by Hummer12007** [comment]
Thanks

**Comment by nbraud** [comment]
Sure

**Comment by nbraud** [comment]
@SirCmpwn The CI reports the build as failed, but has no build log  :/

**Comment by SirCmpwn** [comment]
Resubmitted the build

File: meson.build [hunk 3]
```diff
@@ -19,6 +19,14 @@ is_freebsd = host_machine.system().startswith('freebsd')
 datadir = get_option('datadir')
 sysconfdir = get_option('sysconfdir')
 prefix = get_option('prefix')
+libexecdir = get_option('libexecdir')
+
+if libexecdir == ''
+  libexecdir = join_paths(prefix, 'lib')
+endif
+rundir = join_paths(libexecdir, 'sway')
+add_project_arguments('-DRUNDIR="/@0@"'.format(rundir), language : 'c')
```
Comment (SirCmpwn) [inline meson.build]: Why is this defined as RUNDIR? Just call it LIBEXECDIR everywhere. Be consistent.

File: sway/commands/exec_always.c [hunk 4]
```diff
@@ -1,3 +1,4 @@
+#define _GNU_SOURCE
```
Comment (SirCmpwn) [inline sway/commands/exec_always.c]: no no no no no

Comment (nbraud) [reply sway/commands/exec_always.c]: Aaaaaaaa.  I removed the call to `secure_getent(3)` but forgot to remove that.

File: sway/commands/exec_always.c [hunk 5]
```diff
@@ -51,7 +51,35 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
 	if ((pid = fork()) == 0) {
 		// Fork child process again
 		setsid();
+
 		if ((*child = fork()) == 0) {
+			// Acquire the current PATH
+			char *path = getenv("PATH");
+			const char *extra_path = ":" RUNDIR;
+			const size_t extra_size = sizeof(extra_path);
```
Comment (ascent12) [inline sway/commands/exec_always.c]: sizeof pointer.

File: sway/commands/exec_always.c [hunk 6]
```diff
@@ -51,7 +51,35 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
...
+			} else {
+				size_t n = strlen(path) +1;
+				char *tmp = malloc(n + extra_size);
+				strncpy(tmp, path, n);
+			}
+
+			// Append /usr/lib/sway to PATH
+			strcat(path, extra_path);
+			if (!setenv("PATH", path, 1)) {
+				return cmd_results_new(CMD_FAILURE, "exec_always", "Unable to set PATH");
```
Comment (ascent12) [inline sway/commands/exec_always.c]: `path` is not freed here.

File: sway/commands/exec_always.c [hunk 7]
```diff
@@ -51,7 +51,35 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
...
+				path = malloc(n + extra_size);
+				if (!path) {
+					return cmd_results_new(CMD_FAILURE, "exec_always", "Unable to allocate PATH");
+				}
+				confstr(_CS_PATH, path, n);
+
+			} else {
+				size_t n = strlen(path) +1;
+				char *tmp = malloc(n + extra_size);
+				strncpy(tmp, path, n);
```
Comment (ascent12) [inline sway/commands/exec_always.c]: Copied to a buffer that is immediately discarded (and leaked). You need to add `path = tmp;`.

File: sway/commands/exec_always.c [hunk 8]
```diff
@@ -51,7 +51,35 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
...
+				size_t n = confstr(_CS_PATH, NULL, 0);
+				path = malloc(n + extra_size);
+				if (!path) {
+					return cmd_results_new(CMD_FAILURE, "exec_always", "Unable to allocate PATH");
+				}
+				confstr(_CS_PATH, path, n);
+
+			} else {
+				size_t n = strlen(path) +1;
+				char *tmp = malloc(n + extra_size);
```
Comment (ascent12) [inline sway/commands/exec_always.c]: Missing NULL check.

File: sway/commands/exec_always.c (hunk 7 above)
Comment (nbraud) [reply sway/commands/exec_always.c]: Arg, yes, this was lost during rebase  ;_;
(I just checked, still had it in the `.deb` I built earlier :()

File: meson.build [hunk 9]
```diff
@@ -19,6 +19,14 @@ is_freebsd = host_machine.system().startswith('freebsd')
 datadir = get_option('datadir')
 sysconfdir = get_option('sysconfdir')
 prefix = get_option('prefix')
+libexecdir = get_option('libexecdir')
+
+if libexecdir == ''
+  libexecdir = join_paths(prefix, 'lib')
```
Comment (emersion) [inline meson.build]: Please use tabs instead of spaces

File: sway/commands/exec_always.c [hunk 10]
```diff
@@ -51,7 +51,41 @@ struct cmd_results *cmd_exec_always(int argc, char **argv) {
...
+			if (!path) {
+				size_t n = confstr(_CS_PATH, NULL, 0);
+				path = malloc(n + extra_size);
+				if (!path) {
+					return cmd_results_new(CMD_FAILURE, "exec_always", "Unable to allocate PATH");
+				}
+				confstr(_CS_PATH, path, n);
+
+			} else {
+				size_t n = strlen(path) +1;
```
Comment (emersion) [inline sway/commands/exec_always.c]: Can you add a space between `+` and `1`

File: meson.build [hunk 11]
```diff
@@ -19,6 +19,14 @@ is_freebsd = host_machine.system().startswith('freebsd')
 datadir = get_option('datadir')
 sysconfdir = get_option('sysconfdir')
 prefix = get_option('prefix')
+libexecdir = get_option('libexecdir')
+
+if libexecdir == ''
+	libexecdir = join_paths(prefix, 'lib')
+endif
+rundir = join_paths(libexecdir, 'sway')
+add_project_arguments('-DRUNDIR="/@0@"'.format(rundir), language : 'c')
```
Comment (emersion) [inline meson.build]: Re: https://github.com/swaywm/sway/pull/1881#discussion_r185303178

I also think `RUNDIR` is confusing. Maybe renaming it to `SWAY_LIBEXECDIR` would be better?

Comment (nbraud) [reply meson.build]: Fair; I will do that in a bit  :)

**Comment by nbraud** [comment]
@SirCmpwn @ascent12 @emersion Thanks for the feedback.  Can you confirm this was all handled?

**Comment by emersion** [comment]
Thanks!

**Pull request merged**